atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/AbstractInMemoryTimelockExtension.java

```diff
@@ -0,0 +1,237 @@
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.atlasdb.util.MetricsManagers;
+import com.palantir.conjure.java.api.config.service.PartialServiceConfiguration;
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.serialization.ObjectMappers;
+import com.palantir.lock.LockService;
+import com.palantir.lock.client.CommitTimestampGetter;
+import com.palantir.lock.client.LegacyLeaderTimeGetter;
+import com.palantir.lock.client.LegacyLockTokenUnlocker;
+import com.palantir.lock.client.LockLeaseService;
+import com.palantir.lock.client.NamespacedConjureTimelockServiceImpl;
+import com.palantir.lock.client.RemoteTimelockServiceAdapter;
+import com.palantir.lock.client.RequestBatchersFactory;
+import com.palantir.lock.client.TransactionStarter;
+import com.palantir.lock.v2.NamespacedTimelockRpcClient;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.sls.versions.OrderableSlsVersion;
+import com.palantir.timelock.config.ClusterInstallConfiguration;
+import com.palantir.timelock.config.ImmutableClusterInstallConfiguration;
+import com.palantir.timelock.config.ImmutableDefaultClusterConfiguration;
+import com.palantir.timelock.config.ImmutableSqlitePaxosPersistenceConfiguration;
+import com.palantir.timelock.config.ImmutableTimeLockRuntimeConfiguration;
+import com.palantir.timelock.config.PaxosInstallConfiguration;
+import com.palantir.timelock.config.PaxosInstallConfiguration.PaxosLeaderMode;
+import com.palantir.timelock.config.TimeLockInstallConfiguration;
+import com.palantir.timelock.config.TimeLockRuntimeConfiguration;
+import com.palantir.timestamp.ManagedTimestampService;
+import com.palantir.timestamp.TimestampManagementService;
+import com.palantir.timestamp.TimestampService;
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import org.awaitility.Awaitility;
+
+public abstract class AbstractInMemoryTimelockExtension implements TimeLockServices, Closeable {
+    private static final String USER_AGENT_NAME = "user-agent";
+    private static final String USER_AGENT_VERSION = "3.1415926.5358979";
+    private static final UserAgent USER_AGENT = UserAgent.of(UserAgent.Agent.of(USER_AGENT_NAME, USER_AGENT_VERSION));
+    private static final int THREAD_POOL_SIZE = 128;
+    private static final int BLOCKING_TIMEOUT_MS = 60 * 800; // 0.8 mins to ms
+
+    private String client;
+    private TimeLockAgent timeLockAgent;
+    private TimeLockServices delegate;
+    private TimeLockHelperServices helperServices;
+
+    private LockLeaseService lockLeaseService;
+    private NamespacedConjureTimelockServiceImpl namespacedConjureTimelockService;
+
+    public AbstractInMemoryTimelockExtension() {
```

Comment:
Background info:

We have `InMemoryTimeLockRule` that we can use in our JUnit4 test cases. Now, we need to convert it to an Extension so that we can use it in our JUnit5 test cases. Here is the caveat: With Rules, we can use them as `@Rule` or `@ClassRule` without any additional classes. So, when we have `InMemoryTimeLockRule`, we can use it as a `@Rule` or `@ClassRule`, depending on our test classes. Rules that are annotated with `@Rule` are applied before each test case (similar to BeforeEach), while rules that are annotated with `@ClassRule` are applied before all test cases (similar to BeforeAll).

However, the same thing does not exist for extensions. You need to create two different classes to replace `@ClassRule` (via `BeforeAllCallback` and `AfterAllCallback`)  and another class for `@Rule` (via `BeforeEachCallback` and `AfterEachCallback`). For this reason, I created an abstract class that implements what to do, and I added two subclasses called `InMemoryTimelockExtension` that calls `setup()` and `tearDown()` methods in `BeforeEach()` and `AfterEach()` methods. Additionally, I added a class named `InMemoryTimelockClassExtension` that calls `setup()` and `tearDown()` methods in `BeforeAll()` and `AfterAll()` methods.

Reply:
This class is merged version of `InMemoryTimeLockRule` and `InMemoryTimelockServices`.